### ISSUE TITLE
velocity parsing breaks after minor changes in document

### ISSUE THREAD

**Issue by may-bee** [opensagres/xdocreport#62]
Hi, I am observing this behavior quite for a long time. Now I have some spare time to extract necessary information for analysis. :)
But I need you to analyze the transformations in the document, which seems to break velocity parsing. This is also highly related to #3 as with some tool extracting velocity code from a docx template will help a lot to analyze this issue. Now I have to focus on your experiences and xdocreport knowledge I do not have.

Background: We are maintaining xdocreport templates for generating docx documents. These documents make use of a lot of word features like section breaks, page breaks, text boxes, position frames, and so on... Thus, you can imagine that these documents might have a high potential to create a document.xml, which might has not been considered yet by xdocreport. Sometimes, we observe, that the velocity parsing breaks after changing a sign or a sentence (e.g. here removing a dot), which should not affect the velocity code at all.

I cannot post you the whole document for customer reasons, but maybe we can go through some transformations in the document.xml, which has been performed by word after saving the docx document after removing a dot. Maybe you can make up your mind, whether the shown transformations break the velocity code preprocessing. If not, I can have a look at more strange transformations performed in the document.

**Exception:**
```
Caused by: org.apache.velocity.exception.ParseErrorException: Encountered "#end" at fr.opensagres.xdocreport.document.docx.DocxReport@d822dc!word/document.xml[line 1, column 56068]
Was expecting one of:
    <EOF> 
    "(" ...
    <RPAREN> ...
    <ESCAPE_DIRECTIVE> ...
    <SET_DIRECTIVE> ...
    "##" ...
    "\\\\" ...
    "\\" ...
    <TEXT> ...
    "*#" ...
    "*#" ...
    "]]#" ...
    <STRING_LITERAL> ...
    <IF_DIRECTIVE> ...
    <INTEGER_LITERAL> ...
    <FLOATING_POINT_LITERAL> ...
    <WORD> ...
    <BRACKETED_WORD> ...
    <IDENTIFIER> ...
    <DOT> ...
    "{" ...
    "}" ...
    <EMPTY_INDEX> ...
    
	at org.apache.velocity.Template.process(Template.java:151)
	at org.apache.velocity.runtime.resource.ResourceManagerImpl.loadResource(ResourceManagerImpl.java:437)
	at org.apache.velocity.runtime.resource.ResourceManagerImpl.getResource(ResourceManagerImpl.java:352)
	at org.apache.velocity.runtime.RuntimeInstance.getTemplate(RuntimeInstance.java:1533)
	at org.apache.velocity.app.VelocityEngine.getTemplate(VelocityEngine.java:392)
	at fr.opensagres.xdocreport.template.velocity.internal.VelocityTemplateEngine.processWithCache(VelocityTemplateEngine.java:97)
	at fr.opensagres.xdocreport.template.AbstractTemplateEngine.process(AbstractTemplateEngine.java:111)
	at fr.opensagres.xdocreport.template.AbstractTemplateEngine.process(AbstractTemplateEngine.java:83)
	at fr.opensagres.xdocreport.document.AbstractXDocReport.processTemplateEngine(AbstractXDocReport.java:775)
	at fr.opensagres.xdocreport.document.AbstractXDocReport.process(AbstractXDocReport.java:521)
	at fr.opensagres.xdocreport.document.AbstractXDocReport.process(AbstractXDocReport.java:487)
	at de.bund.bva.fachverfahrensfactory.common.schreiben.core.schreiben.impl.SchreibenVerwalterImpl.erzeugeSchreiben(SchreibenVerwalterImpl.java:131)
	... 34 more
```

**snippet from valid document:**
```xml
			<w:fldSimple w:instr=" MERGEFIELD  &quot;#foreach($i in [0..$SammlungUtil.anzahlSammlungenRelativZuNull])&quot;  \* MERGEFORMAT ">
				<w:r w:rsidR="00957B95">
					<w:rPr>
						<w:rFonts w:cs="Arial"/>
						<w:noProof/>
						<w:sz w:val="18"/>
						<w:szCs w:val="22"/>
					</w:rPr>
					<w:t>«#foreach($i in [0..$SammlungUtil.anzahlS»</w:t>
				</w:r>
			</w:fldSimple>
			<w:fldSimple w:instr=" MERGEFIELD  #set($adressat=$SammlungUtil.objektZuSammlung($i,'adressat'))#set($pruefungsausschuss=$SammlungUtil.objektZuSammlung($i,'pruefungsausschuss'))#set($pruefung=$SammlungUtil.objektZuSammlung($i,'pruefung'))  \* MERGEFORMAT ">
				<w:r w:rsidR="00141997">
					<w:rPr>
						<w:rFonts w:cs="Arial"/>
						<w:noProof/>
						<w:sz w:val="18"/>
						<w:szCs w:val="22"/>
					</w:rPr>
					<w:t>«#set($adressat=$SammlungUtil.objektZuSam»</w:t>
				</w:r>
			</w:fldSimple>
			<w:fldSimple w:instr=" MERGEFIELD  #set($pruefungsausschuss=$SammlungUtil.objektZuSammlung($i,'pruefungsausschuss'))#set($funktionPaMitglied=$SammlungUtil.objektZuSammlung($i,'pruefungsausschussmitglied.normal'))  \* MERGEFORMAT ">
				<w:r w:rsidR="00A94E54" w:rsidRPr="00A94E54">
					<w:rPr>
						<w:rFonts w:cs="Arial"/>
						<w:noProof/>
						<w:sz w:val="18"/>
						<w:szCs w:val="22"/>
					</w:rPr>
					<w:t>«#set($pruefungsausschuss=$SammlungUtil.o»</w:t>
				</w:r>
			</w:fldSimple>
```

**snippet from invalid document:**
```xml
			<w:r w:rsidR="00AE6684">
				<w:fldChar w:fldCharType="begin"/>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:instrText xml:space="preserve"> MERGEFIELD  "#foreach($i in [0..$SammlungUtil.anzahlSammlungenRelativZuNull])"  \* MERGEFORMAT </w:instrText>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:fldChar w:fldCharType="separate"/>
			</w:r>
			<w:r w:rsidR="00957B95">
				<w:rPr>
					<w:rFonts w:cs="Arial"/>
					<w:noProof/>
					<w:sz w:val="18"/>
					<w:szCs w:val="22"/>
				</w:rPr>
				<w:t>«#foreach($i in [0..$SammlungUtil.anzahlS»</w:t>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:rPr>
					<w:rFonts w:cs="Arial"/>
					<w:noProof/>
					<w:sz w:val="18"/>
					<w:szCs w:val="22"/>
				</w:rPr>
				<w:fldChar w:fldCharType="end"/>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:fldChar w:fldCharType="begin"/>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:instrText xml:space="preserve"> MERGEFIELD  #set($adressat=$SammlungUtil.objektZuSammlung($i,'adressat'))#set($pruefungsausschuss=$SammlungUtil.ob</w:instrText>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:instrText xml:space="preserve">jektZuSammlung($i,'pruefungsausschuss'))#set($pruefung=$SammlungUtil.objektZuSammlung($i,'pruefung'))  \* MERGEFORMAT </w:instrText>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:fldChar w:fldCharType="separate"/>
			</w:r>
			<w:r w:rsidR="00141997">
				<w:rPr>
					<w:rFonts w:cs="Arial"/>
					<w:noProof/>
					<w:sz w:val="18"/>
					<w:szCs w:val="22"/>
				</w:rPr>
				<w:t>«#set($adressat=$SammlungUtil.objektZuSam»</w:t>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:rPr>
					<w:rFonts w:cs="Arial"/>
					<w:noProof/>
					<w:sz w:val="18"/>
					<w:szCs w:val="22"/>
				</w:rPr>
				<w:fldChar w:fldCharType="end"/>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:fldChar w:fldCharType="begin"/>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:instrText xml:space="preserve"> MERGEFIELD  #set($pruefungsausschuss=$SammlungUtil.objektZuSammlung($i,'pruefungsausschuss')</w:instrText>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:instrText xml:space="preserve">)#set($funktionPaMitglied=$SammlungUtil.objektZuSammlung($i,'pruefungsausschussmitglied.normal'))  \* MERGEFORMAT </w:instrText>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:fldChar w:fldCharType="separate"/>
			</w:r>
			<w:r w:rsidR="00A94E54" w:rsidRPr="00A94E54">
				<w:rPr>
					<w:rFonts w:cs="Arial"/>
					<w:noProof/>
					<w:sz w:val="18"/>
					<w:szCs w:val="22"/>
				</w:rPr>
				<w:t>«#set($pruefungsausschuss=$SammlungUtil.o»</w:t>
			</w:r>
			<w:r w:rsidR="00AE6684">
				<w:rPr>
					<w:rFonts w:cs="Arial"/>
					<w:noProof/>
					<w:sz w:val="18"/>
					<w:szCs w:val="22"/>
				</w:rPr>
				<w:fldChar w:fldCharType="end"/>
			</w:r>
```

There are more changes performed by word in the document.xml. Nevertheless, I do not want to post them all directly. If you cannot see anything unusal, I will search for more strange reformattings :). Hope you will have some time to look at this issue. Thanks in advance!

**Comment by may-bee** [comment]
Additionally, the location named in the exception points at the location behind the following fragment `<w:highlight w:va` in

```xml
<w:r w:rsidR="004973CC" w:rsidRPr="007A3417">
				<w:rPr>
					<w:highlight w:val="yellow"/>
				</w:rPr>
				<w:t>__________________________</w:t>
			</w:r>
```

all changes:
snippet from valid document:
```xml
			<w:fldSimple w:instr=" MERGEFIELD  $!ausbildungsberufMaennlich  \* MERGEFORMAT ">
				<w:r w:rsidR="00456A64">
					<w:rPr>
						<w:noProof/>
					</w:rPr>
					<w:t>«$!ausbildungsberufMaennlich»</w:t>
				</w:r>
			</w:fldSimple>
			<w:fldSimple w:instr=" MERGEFIELD  #else  \* MERGEFORMAT ">
				<w:r w:rsidR="004973CC" w:rsidRPr="005B1410">
					<w:t>«#else»</w:t>
				</w:r>
			</w:fldSimple>
			<w:r w:rsidR="004973CC" w:rsidRPr="007A3417">
				<w:rPr>
					<w:highlight w:val="yellow"/>
				</w:rPr>
				<w:t>__________________________</w:t>
			</w:r>
			<w:fldSimple w:instr=" MERGEFIELD  #end  \* MERGEFORMAT ">
				<w:r w:rsidR="004973CC" w:rsidRPr="005B1410">
					<w:t>«#end»</w:t>
				</w:r>
			</w:fldSimple>
			<w:r w:rsidR="00140EE6">
				<w:t xml:space="preserve"> </w:t>
			</w:r>
			<w:r w:rsidR="00F9771F">
				<w:t>/</w:t>
			</w:r>
			<w:r w:rsidR="00140EE6">
				<w:t xml:space="preserve"> </w:t>
			</w:r>
			<w:fldSimple w:instr=" MERGEFIELD  #set($ausbildungsberufWeiblich=$!SvzUtil.liesBerufOhneFachrichtungGeschlechtspezifisch($pruefungsausschuss.ausbildungsberufOhneFachrichtung,\&quot;W\&quot;))  \* MERGEFORMAT ">
				<w:r w:rsidR="00456A64">
					<w:rPr>
						<w:noProof/>
					</w:rPr>
					<w:t>«#set($ausbildungsberufWeiblich=$!SvzUtil»</w:t>
				</w:r>
			</w:fldSimple>
```

*snippet from invalid document:*
```xml
			<w:r>
				<w:fldChar w:fldCharType="begin"/>
			</w:r>
			<w:r>
				<w:instrText xml:space="preserve"> MERGEFIELD  $!ausbildungsberufMaennlich  \* MERGEFORMAT </w:instrText>
			</w:r>
			<w:r>
				<w:fldChar w:fldCharType="separate"/>
			</w:r>
			<w:r w:rsidR="00456A64">
				<w:rPr>
					<w:noProof/>
				</w:rPr>
				<w:t>«$!ausbildungsberufMaennlich»</w:t>
			</w:r>
			<w:r>
				<w:rPr>
					<w:noProof/>
				</w:rPr>
				<w:fldChar w:fldCharType="end"/>
			</w:r>
			<w:r>
				<w:fldChar w:fldCharType="begin"/>
			</w:r>
			<w:r>
				<w:instrText xml:space="preserve"> MERGEFIELD  #else  \* MERGEFORMAT </w:instrText>
			</w:r>
			<w:r>
				<w:fldChar w:fldCharType="separate"/>
			</w:r>
			<w:r w:rsidR="004973CC" w:rsidRPr="005B1410">
				<w:t>«#else»</w:t>
			</w:r>
			<w:r>
				<w:fldChar w:fldCharType="end"/>
			</w:r>
			<w:r w:rsidR="004973CC" w:rsidRPr="007A3417">
				<w:rPr>
					<w:highlight w:val="yellow"/>
				</w:rPr>
				<w:t>__________________________</w:t>
			</w:r>
			<w:r>
				<w:fldChar w:fldCharType="begin"/>
			</w:r>
			<w:r>
				<w:instrText xml:space="preserve"> MERGEFIELD  #end  \* MERG</w:instrText>
			</w:r>
			<w:r>
				<w:instrText xml:space="preserve">EFORMAT </w:instrText>
			</w:r>
			<w:r>
				<w:fldChar w:fldCharType="separate"/>
			</w:r>
			<w:r w:rsidR="004973CC" w:rsidRPr="005B1410">
				<w:t>«#end»</w:t>
			</w:r>
			<w:r>
				<w:fldChar w:fldCharType="end"/>
			</w:r>
			<w:r w:rsidR="00140EE6">
				<w:t xml:space="preserve"> </w:t>
			</w:r>
			<w:r w:rsidR="00F9771F">
				<w:t>/</w:t>
			</w:r>
			<w:r w:rsidR="00140EE6">
				<w:t xml:space="preserve"> </w:t>
			</w:r>
			<w:r>
				<w:fldChar w:fldCharType="begin"/>
			</w:r>
			<w:r>
				<w:instrText xml:space="preserve"> MERGEFIELD  #set($ausbildungsberufWeiblich=$!SvzUtil.liesBerufOhneFachrichtungGeschlechtspezifisch($pruefungsausschuss.ausbildungsberufOhneFachrichtung,\"W\"))  \* MERGEFORMAT </w:instrText>
			</w:r>
			<w:r>
				<w:fldChar w:fldCharType="separate"/>
			</w:r>
			<w:r w:rsidR="00456A64">
				<w:rPr>
					<w:noProof/>
				</w:rPr>
				<w:t>«#set($ausbildungsberufWeiblich=$!SvzUtil»</w:t>
			</w:r>
			<w:r>
				<w:rPr>
					<w:noProof/>
				</w:rPr>
				<w:fldChar w:fldCharType="end"/>
			</w:r>
```

**Comment by angelozerr** [comment]
@may-bee thanks for your info, but as we develop XDocReport in our spare time, we cannot try and guess each issues. So if you wish we help you, please provide us (create a github project) a simple sample docx+java main + context which causes your problem. For docx I suggest you that you remove the whole content of your customer to have a simple docx which causes your problem.

**Comment by may-bee** [comment]
> For docx I suggest you that you remove the whole content of your customer to have a simple docx which causes your problem.

might get a little bit freaky as word changes the document's xml structure at each saving... so no idea whether the same problem will also hold, when I start to delete all the contents. But I can try so.
So at the above snippets you cannot see any xml structure, you might have overlooked which may cause the issue?

**Comment by angelozerr** [comment]
> might get a little bit freaky as word changes the document's xml structure at each saving...

Don't use MS Word. Unzip docx, updat ethe document.xml at hand.

**Comment by kevinsdooapp** [comment]
@may-bee https://github.com/opensagres/xdocreport/issues/77 could be the same problem... can you verify that the procedure in issue77 fix your problem? If yes, this issue is related to the same problem.

**Comment by may-bee** [comment]
I observed, that under some circumstances not all merge fields are detected as velocity code. The document, I edited was invalid before editing, but could be processed by velocity anyhow.
Right after opening, editing, and saving the document, the document becomes invalid for velocity processing. Your issue might be a good explanation to that.

I would suggest to close this issue as it remains open what exaclty is the cause of velocity or xdocreport to skip any control statements in a document. For now I would follow the idea of issue #77 and suspend analysis here.

**Comment by kevinsdooapp** [comment]
@may-bee could you provide us your template that failed please? I'm looking for a smaller template than mine.

Thank you

**Comment by angelozerr** [comment]
> could you provide us your template that failed please?

template + Pojo + simple Java main which causes the problem please.

**Comment by may-bee** [comment]
sry for being quite late here, I missed this responsed due to lack of spare time :)
I will have a look, whether I can provide you the template, but without great confidence as I deleted a lot old data the last month.
As I could solve my issue by correcting the template, it was fine for me no matter the question has left open why an invalid template could have been parsed successfully.

**Comment by hoangductrung63** [comment]
@angelozerr : i have same issue as @maybeec. My docx attached.
[BDS 03 BEN-BEN VAY KHDN BEN BAO DAM KHDN.docx](https://github.com/opensagres/xdocreport/files/573017/BDS.03.BEN-BEN.VAY.KHDN.BEN.BAO.DAM.KHDN.docx)

Exception: org.apache.velocity.exception.ParseErrorException: Encountered "<EOF>" at reportID!word/document.xml[line 1, column 3704]
Was expecting one of:
    "(" ...
    <RPAREN> ...
    <ESCAPE_DIRECTIVE> ...
    <SET_DIRECTIVE> ...
    "##" ...
    "\\\\" ...
    "\\" ...
    <TEXT> ...
    "*#" ...
    "*#" ...
    "]]#" ...
    <STRING_LITERAL> ...
    <END> ...
    <IF_DIRECTIVE> ...
    <INTEGER_LITERAL> ...
    <FLOATING_POINT_LITERAL> ...
    <WORD> ...
    <BRACKETED_WORD> ...
    <IDENTIFIER> ...
    <DOT> ...
    "{" ...
    "}" ...
    <EMPTY_INDEX> ...

My solution : i must hand job delete «#foreach($MBTSBD03_L2 in $MBTSBD03_L2s)» and Ctrl F9, add new merged field #foreach($MBTSBD03_L2 in $MBTSBD03_L2s).
Please help me understand root of problem !

**Comment by hoangductrung63** [comment]
Exception: org.apache.velocity.exception.ParseErrorException: Encountered "<EOF>" at reportID!word/document.xml[line 1, column 3704]
Was expecting one of:
    "(" ...
    <RPAREN> ...
    <ESCAPE_DIRECTIVE> ...
    <SET_DIRECTIVE> ...
    "##" ...
    "\\\\" ...
    "\\" ...
    <TEXT> ...
    "*#" ...
    "*#" ...
    "]]#" ...
    <STRING_LITERAL> ...
    <END> ...
    <IF_DIRECTIVE> ...
    <INTEGER_LITERAL> ...
    <FLOATING_POINT_LITERAL> ...
    <WORD> ...
    <BRACKETED_WORD> ...
    <IDENTIFIER> ...
    <DOT> ...
    "{" ...
    "}" ...
    <EMPTY_INDEX> ...

**Comment by rutvikupadhyay** [comment]
@angelozerr : I am also getting similar stack trace as @hoangductrung63 has got. I have the the document but its quite large so difficult to figure out where the problem is. Though when the same document's content is copy-paste in to new word document & save, it doesn't throw any error after that. Any idea/thoughts?

I am using xDocReport 1.0.2 release & MS Office 2007.

**Comment by hoangductrung63** [comment]
@rutvikupadhyay : Exactly, when copy/paste , it 's has a problem. I don't know why, but it 's happen when use Office 2007, let's try Office 2013.
Other topic related : [https://github.com/opensagres/xdocreport/issues/172](url)

**Comment by angelozerr** [comment]
>  I have the the document but its quite large so difficult to figure out where the problem is.

Without simple docx, I cannot help you.

> I am using xDocReport 1.0.2

Wow, 1.0.2 was released from 2013. Please try with 1.0.6 we have fixed a lot of bugs since 2013.

**Comment by rutvikupadhyay** [comment]
@angelozerr I understand your concern but unfortunately I don't have the one page document. I'll let you know once available to me. Is 1.0.6 ready as I can see only 1.0.5 as the latest release on the wiki page. Further more, no change at all in the docx template (add & remove a space somewhere in the doc) cause velocity parsing exception which was working prior to change. Though I noted size difference between document.xml files created for the docx files.

I also found similar issue https://github.com/opensagres/xdocreport/issues/77 & as per your comment in it, you did commit some code. Did you fix any specific issue or general parsing issue? Please advise.

**Comment by angelozerr** [comment]
> I understand your concern but unfortunately I don't have the one page document.

I cannot help you if I have not a little docx.

> I can see only 1.0.5 as the latest release on the wiki page. 

It's hard for us to find time. To know the last version you can see at https://github.com/opensagres/xdocreport with "maven central icon" [![Maven Central](https://img.shields.io/maven-central/v/fr.opensagres.xdocreport/xdocreport.svg)](http://search.maven.org/#search%7Cga%7C1%7Cg%3A%22fr.opensagres.xdocreport%22%20AND%20a%3A%22xdocreport%22)

Any contribution are welcome!

> I also found similar issue #77 

The Velocity error parsing can be thrown for any reasons (template is not valid because you writ enot correctly or because it's a XDocReport bug which process not correctly the word/document.xml)

I suggest really that you test with 1.0.6 which fixes a lot of bugs of preprocessing (1.0.2 is so old).

**Comment by rutvikupadhyay** [comment]
@angelozerr , thanks for the input. It worked perfectly for the templates. Though I have one question. The new jar has broken the code which used to work previously. java.lang.NoSuchMethodError: fr.opensagres.xdocreport.document.images.SimpleImageInfo.(java/io/File;). Is there any other way to achieve the same? Thanks
